build/bundle.js

```diff
@@ -116,121 +78,135 @@ class MainPage {
-     * 추가 된 라우트를 확인하고 이동하는 함수.
-     * 쿼리파라미터가 존재했을때 URL에 합쳐서 경로를 넘겨준다. 
-     * URL에서 쿼리스트링 값을 체크하고 저장한다.
-     * * @param {object} hashRouterPages 
-     */
-    checkRoutes(){
-      window.onhashchange = () => {
-        this.hash = window.location.hash;
-        this.setQueryString()
-        this.setQueryParameter()
-        this.addRoute(this.hashRouterPages);
-      };
-    }
-    
-    /**
-     * 모든 라우터에서 일치한 하나의 페이지를 추가하는 함수
-     * 라우터 등록되지 않은 페이지로 이동했을때 404 페이지로 이동.
-     * @param {Page} hashRouterPages 
-     */
-    addRoute(hashRouterPages){
-      const findPage = hashRouterPages.find((page) => page.toPath === this.hash);
+    })
+  }
+
+  /**
+   * URL에서 해쉬 값을 체크 하고 저장한다.
+   * 쿼리파라미터가 존재했을때 URL에 합쳐서 경로를 넘겨준다. 
+   * URL에서 쿼리스트링 값을 체크하고 저장한다.
+   * 모든 라우트에서 일치된 라우트를 확인하고 해당 페이지로 이동하는 함수.
+   */
+  router.checkRoutes = () => {
+    window.onhashchange = () => {
+      hash = window.location.hash;
+      setQueryString()
+      setQueryParameter()
+      
+      const findPage = hashRouterPages.find((page) => page.toPath === hash);
+
       if(findPage){
         const ViewPage = findPage.page;
         this.currentPage = new ViewPage({ router: this });
-        this.app.innerHTML += this.currentPage.render();
+        app.innerHTML += this.currentPage.render();
       }else{
-        const NotFoundPage = hashRouterPages.find((page) => page.toPath === this.notFoundPage).page;
+        const NotFoundPage = hashRouterPages.find((page) => page.toPath === notFoundPage).page;
         this.currentPage = new NotFoundPage({ router: this });
-        this.push(this.notFoundPage)
+        navigate(this.notFoundPage)
       }
+    };
+  }
+
+   /**
+   * 지정된 이름으로 이동하는 함수.
+   * @param {string} pageName 
+   */
+  router.navigate = (pageName) => {
+    app.innerHTML = '';
+    window.location.hash = checkHashUrl(pageName);
+  }
+
+  /**
+   * 라우터를 등록한다. 
+   */
+  router.addRouter = (item) => {
+    hashRouterPages.push(item)
+  }
+
+  /**
+   * 404 페이지를 설정한다.
+   * @param {Page} page 
+   */
+  router.setNotFound = (page) => {
+    notFoundPage = page.path;
+  }
+   
+  /**
+   * 전달받은 url 에 대한 한글 디코딩을 실행한다.
+   * 전달받은 값에 대한 공백 제거를 실행한다. 
+   * queryString 존재 시 url 에 set.
+   * parameter 존재 시 url 에 set.
+   * @param {String} page 
+   * @returns 
+   */
+    const checkHashUrl =(page) => {
+    let url = page;
+
+    if(this.query) url = page + this.query
+    if(this.parameter) url = page + this.parameter
+
+    const decodeURL = decodeURI(url).replace(/ /g, '')
+    return decodeURL;
+  } 
+
+  /**
+   * URL 에 쿼리스트링이 있을시 set 한다.
+   */
+  const setQueryString= () => {
+    this.query = null;
+    const queryStringIndex = window.location.hash.indexOf('?')
+    if(queryStringIndex > 0){
+      hash = window.location.hash.slice(0, queryStringIndex);
+      const queryStringUrl = window.location.hash.slice(queryStringIndex);
+      const url = new URLSearchParams(queryStringUrl);
+      this.query = '?'+url.toString();
```

Comment:
요즘은 이렇게 많이 쓰는 것 같아요 ㅎㅎ

```suggestion
      this.query = `?${url}`
```